### DISCUSSION TITLE
Airflow triggering old code

### DISCUSSION THREAD

**Discussion by tikkisetti** [apache/airflow#24258] [Q&A]
### Apache Airflow version

2.1.4

### What happened

Some of our dags are still running the old code. One way to get them pick the latest code is to delete the dag and let it reload.  How do I fix this? 

### Operating System

Debian GNU/Linux 10

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
I tihnk if you need help you must provide some details, logs, circumstances, cionfiguration, the way you run Airflow etc. 

Otherwise you make very difficult for others who even would lke to help you.

**Reply by OliverLeighC** [reply]
We run airflow in k8 using the official helm chart and a docker image with all the dags and scripts. 

I have exec'd into the individual pods and confirmed that they all have the updated code (the python scripts that the dags are running), however when the dag actually runs it is running an older version of the script from a couple weeks ago. I was wondering if there was some caching doing on since all the pods have the up to date code.

**Reply by potiuk** [reply]
It runs whatever the DAGs are in the DAGS_FOLDER. The only caching that might happen is if you have python bytecode *.pyc and __pycache__ files. This is the usual "python" caching that might happen. Deleting those might help, it's just interesting why they are not refreshed by Python automatically. One of the scenarios I can imagine - It could be that time on one of your system that either reads or wirte the dags has been awfully wrong and that some of those files have future date (and Python won't delete them). 

Another scenario is (you did not explain the way how you sync your dags or how kind of structure of your modules you have ) that the usual python import rules kick-in and that older version of the code imported by your DAGs is available on PYTHONPATH before the newer ones - this is standard python behaviour, and depends on details on your deployment configuration and how you distribute the DAGS. This is your deploiyment detail, but for now I can only guess - if you want more help, you need to (as I explained before) explain details of your deployment.

**Reply by potiuk** [reply]
Please help us to help you if you need our help in solving your problem.

**Reply by OliverLeighC** [reply]
We use codeship to push the image to AWS ECR, then the new image gets picked up by argocd which updates the image tag on the k8 cluster. 

Here is the Dockerfile used to build the image. I checked the code running in the kubernetes pods and the dags folder has the correct up to date code. 

```
FROM apache/airflow:2.1.4-python3.8
USER root
ARG AIRFLOW_USER_HOME=/opt/airflow

RUN mkdir -p ${AIRFLOW_USER_HOME}/temp

# MSSQL driver installation pod
RUN set -ex \
    && whoami \
    && apt-get update \
    && apt-get -y install sudo \
    && apt-get install -y gnupg2 \
    && apt-get install -y curl \
    && apt-get install -y unzip \
    && curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add - \
    && curl https://packages.microsoft.com/config/debian/10/prod.list > /etc/apt/sources.list.d/mssql-release.list \
    && apt-get update \
    && ACCEPT_EULA=Y apt-get install -y msodbcsql17 \
    && ACCEPT_EULA=Y apt-get install -y mssql-tools \
    && echo 'export PATH="$PATH:/opt/mssql-tools/bin"' >> ~/.bash_profile \
    && echo 'export PATH="$PATH:/opt/mssql-tools/bin"' >> ~/.bashrc \
    && . ~/.bashrc \
    && apt-get install -y unixodbc-dev \
    && apt-get install -y libgssapi-krb5-2 \
    && apt-get install -y g++

RUN curl -sL https://deb.nodesource.com/setup_12.x -o nodesource_setup.sh \
    && curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - \
    && echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list \
    && bash ./nodesource_setup.sh && apt update && apt install -y nodejs && apt install yarn

RUN npm install -g typescript && npm install -g ts-node

RUN curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64-2.0.30.zip" -o "awscliv2.zip" \
    && unzip awscliv2.zip \
    && ./aws/install

RUN chown -R airflow: ${AIRFLOW_USER_HOME}

USER airflow

COPY . .

RUN pip install apache-airflow[mssql,slack,s3,odbc]==2.1.4
RUN pip install -r ./requirements.txt

USER root
RUN ["ls", "-lrt", "./run_npm.sh"]
#RUN ["chmod", "+x", "./run_npm.sh"]
RUN ./run_npm.sh

#ENTRYPOINT ["/entrypoint.sh"]

RUN apt-get -y install git
#SSH to Github
ARG GITHUB_KEY

ENV an_env_var=$GITHUB_KEY

RUN touch github_key && echo $GITHUB_KEY | base64 --decode > github_key && cat github_key

COPY package.json package.json
COPY package-lock.json package-lock.json

RUN chmod 400 github_key && eval $(ssh-agent) && \
  ssh-add github_key && \
  ssh-keyscan -H github.com >> /etc/ssh/ssh_known_hosts && \
  yarn config set network-timeout 60000 -g && yarn install

USER airflow
```

Our directory structure has a `dags` folder in root, and then the python scripts run by the dags live in `/dags/src/scripts`.

**Reply by OliverLeighC** [reply]
the helm chart looks like (we use a helm dependency because of some weirdness with how argocd handles custom values files for 3rd party helm charts)
```
apiVersion: v1
kind: application
description: Chart for airflow dependency
name: airflow
version: 1.0.0
dependencies:
  - name: airflow
    repository: https://airflow-helm.github.io/charts
    version: 8.5.3
```
and then in the values.yml file `airflow.airflow.image` points to the latest ecr image

**Reply by OliverLeighC** [reply]
one of the dags that is having the problem looks like this. The ecr image running in the pods has the latest version of `src.scripts.secondary_assignments` but when the dag actually runs it is running an older version of that file

```
from datetime import timedelta
import pendulum
from airflow import DAG
from airflow.operators.python import PythonOperator

from src.scripts.secondary_assignments import runScript
from src.utils.common_utils import task_fail_slack_alert, get_variable

variable_config = get_variable("global_config")
env = variable_config["environment"]
dag_schedule = "0 0 * * 1"
mssql_conn_id = "mssql_db"

default_args = {'owner': 'airflow',
                'start_date': pendulum.datetime(2020, 4, 29, tz="US/Eastern"),
                'retries': 2,
                'retry_delay': timedelta(minutes=5),
                'on_failure_callback': task_fail_slack_alert
                }

dag = DAG('Update_Secondary_Assignments',
          schedule_interval=dag_schedule,
          catchup=False,
          default_args=default_args)

update_assignments = PythonOperator(
    task_id='update_assignments',
    python_callable=runScript,
    op_args=[mssql_conn_id, env],
    dag=dag
)

update_assignments
```

**Reply by OliverLeighC** [reply]
Let me know if you need any other info

**Reply by potiuk** [reply]
You have not written how you distribute your DAGs - they do not seem to be embedded in the image if I see well.


Honestly.  I cannot help you too much - and even less give you ready answers. 
This MUST be something specific to your setup. 

You need to dig deeper - exec into the running container and see what is your pythonpath and where the file you got actually  imported from. Search in the container ALL disks and find where `secondary_assignments.py` are `secondary_assignments.pyc` can be found (see below). There is simply NO WAY that your POD executes a code that is not present on the disk of the POD. Airlfow does not do any caching of the code, it solely relies on default mechanisms which come with Python.

Speakiong of which - wild guess one of the possibilities is that somewhere in your system you have stale `secondary_assignments.pyc` file which contains python bytecode from the older version and for whatever reasons (no permissions? time difference?) the usual python mechanims to refresh the bytecode fail - and you run the old bytecode. But where and how and why they are not refreshed, only you can tell. It could be part of your image, or maybe your distributed volume has wrong time/date set and the .pyc files seem to be newer than the .py code for example, or maybe the .pyc files have read-only permission.

If this is the case then delete the .pyc file and set PYTHONDONTWRITEBYTECODE="1" as default value everywhere in your environment if you are not able to find and remove the root cause for the problem. This will prevent python from creating .pyc files. It's not really a disadvantage in K8S world - if you are using K8S executor and your pods are created from the scratch, the bytecode files are not used anyway (they are only useful if a process would be restarted in the same POD after the files are created.

I hope this is something along the lines.

**Comment by voycey** [comment]
We are having the same issue, no amount of deleting pyc works and setting PYTHONDONTWRITEBYTECODE causes everything to crash.
We are running this on EC2 using Docker Compose, the UI and the DAGS seem to flick between new and old versions of the code.
We are seeing this across multiple unrelated projects too.

**Reply by potiuk** [reply]
You likely have somewhere a left-over old Airflow version running. I saw that happening, if your UI is flickering like that. Take a look at your deployment and containers running. There is nothing Airflow can do when you have some old process using old DAGs using the same database running at the same time.

**Reply by voycey** [reply]
We identified that something else was running - I will get one of my colleagues to detail what it was but by looking into the `serialized_dag` table it helped us track down what was happening!

**Reply by aakashnand** [reply]
Thanks @voycey and @potiuk 

I have found out the root cause of this. One more file in DAG folder was using same task name and same DAG_ID which was conflicting. 

This conflict file was left for development purpose. Now the issue is resolved

**Reply by Venture200** [reply]
@aakashnand THANK YOU! You're the savior! I am glad that I came across your comment!
